### PR TITLE
WIP Added recent list feature

### LINKED ISSUE #49: Add call history
The design is TBD. Having a separate drawer is a potential candidate.

### PR DESCRIPTION
**Pull request by akshitkrnagpal**
Resolves #49.

TODO
- [x] Convert all `px` to `em`
- [x] Make changes in grid. (Ongoing)
- [x] Create actions `CONFERENCE_JOINED` and `CONFERENCE_ENDED` and use middleware.
- [x] Declare type Conference.
- [x] Format Date/Time and duration. (using Moment.js)
- [x] Hide Server URL protocol in cards
- [x] Persist recent list in redux.

### CODE REVIEW

**Comment by akshitkrnagpal** [comment]
Ping @saghul.
PTAL.

File: app/features/welcome/styled/Header.js [hunk 1]
```diff
@@ -6,5 +6,5 @@ export default styled.div`
     align-items: center;
     display: flex;
     margin: 0 auto;
-    padding: 30px;
+    padding: 8em;
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/welcome/styled/Header.js]: I was using `px` because we didn't have much text before.
But since we have a little text I thought let's switch to `em`. I'll change all of them to `em`.

File: app/features/welcome/components/Welcome.js [hunk 2]
```diff
@@ -11,9 +11,15 @@ import { connect } from 'react-redux';
...
 import { Navbar } from '../../navbar';
+import { RecentList } from '../../recent-list';
 import { normalizeServerURL } from '../../utils';
 
-import { WelcomeWrapper as Wrapper, Content, Form } from '../styled';
+import {
+    Body,
+    Form,
+    Header,
+    WelcomeWrapper as Wrapper
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/welcome/components/Welcome.js]: Yeah. I'll do that

File: app/features/recent-list/styled/RecentListContainer.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// @flow
+
+import styled from 'styled-components';
+
+export default styled.div`
+    display: grid;
+    grid-template-columns: repeat(4, 25%);
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/recent-list/styled/RecentListContainer.js]: Hmm. I'll make it responsive. Because 3 looked very wide when I tested on large screen.

File: app/features/recent-list/reducer.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+function _insertConference(recentList: Array<Object>, newConference: Object) {
+
+    // Add start time to conference.
+    newConference.startTime = Date.now();
+
+    // Remove same conference.
+    // $FlowFixMe
+    const newRecentList = recentList.filter(
+        (conference: Object) => conference.room !== newConference.room
+            || conference.serverURL !== newConference.serverURL);
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/recent-list/reducer.js]: Yeah. My Bad. :see_no_evil:

File: app/features/recent-list/components/RecentList.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    /**
+     * Redux dispatch.
+     */
+    dispatch: Dispatch<*>;
+
+    /**
+     * Array of recent conferences.
+     */
+    _recentList: Array<Object>;
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/recent-list/components/RecentList.js]: Cool. `type Conference` coming soon.

File: app/features/conference/components/Conference.js [hunk 6]
```diff
@@ -239,11 +240,19 @@ class Conference extends Component<Props, State> {
         setupAlwaysOnTopRender(this._api);
         setupWiFiStats(iframe);
 
-        this._api.on('readyToClose', (event: Event) =>
-            this._navigateToHome(event));
+        this._api.on('readyToClose', (event: Event) => {
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/conference/components/Conference.js]: Sounds great. I'll do that.

File: app/features/welcome/styled/Header.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (saghul) [reply app/features/welcome/styled/Header.js]: Thanks!

File: app/features/recent-list/styled/RecentListContainer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (saghul) [reply app/features/recent-list/styled/RecentListContainer.js]: Let's also leave some space of the sides, so it looks a bit less busy.

File: app/features/recent-list/components/RecentList.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (saghul) [reply app/features/recent-list/components/RecentList.js]: Since this is a recent list, how about some `RecentListItem` instead? We don't really have a `Conference` type, and adding it for this sounds a bit odd to me.

File: app/features/recent-list/reducer.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/recent-list/reducer.js]: Nops. It is `||`.

We want to remove exact same conference from array, ie,
```
conference.room === newConference.room && conference.serverURL === conference.serverURL
```
should be true.

Therefore we'll have to pass the inverse of the above to the filter function.
Which will be 
```
conference.room !== newConference.room || conference.serverURL !== newConference.serverURL
```

File: app/features/recent-list/styled/RecentListContainer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (saghul) [reply app/features/recent-list/styled/RecentListContainer.js]: Not sure. Let's play with it anyway, changing it shouldn't be a big deal, right?

File: package.json [hunk 7]
```diff
@@ -93,6 +93,7 @@
     "history": "4.7.2",
     "jitsi-meet-electron-utils": "github:jitsi/jitsi-meet-electron-utils#1972c3bf0884ace68eb496894dabae593d6dbf49",
     "js-utils": "github:jitsi/js-utils#0c53500a5120be2aa3fc590f0f932a0d4771920f",
+    "moment": "^2.22.2",
```
Comment (saghul) [inline package.json]: Please ping the dependency.

File: app/features/welcome/styled/Form.js [hunk 8]
```diff
@@ -3,6 +3,6 @@
 import styled from 'styled-components';
 
 export default styled.form`
-    width: 350px;
-    margin: 0 15px;
+    width: 26em;
```
Comment (saghul) [inline app/features/welcome/styled/Form.js]: Can you please make these changes a separate commit?

File: package.json (hunk 7 above)
Comment (akshitkrnagpal) [reply package.json]: Oops.

File: app/features/welcome/styled/Form.js (hunk 8 above)
Comment (akshitkrnagpal) [reply app/features/welcome/styled/Form.js]: Sure. I'll separate them after we finalize everything else.

**Comment by akshitkrnagpal** [comment]
screen shot 2018-07-24 at 5 35 08 pm

File: app/features/recent-list/actionTypes.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+export const SAVE_CURRENT_CONFERENCE = Symbol('SAVE_CURRENT_CONFERENCE');
+
+/**
+ * The type of (redux) action that updates the endtime of the last conference.
+ *
+ * {
+ *     type: UPDATE_END_TIME
+ * }
+ */
+export const UPDATE_END_TIME = Symbol('UPDATE_END_TIME');
```
[the remark this answers is not in the document]
Comment (saghul) [reply app/features/recent-list/actionTypes.js]: Because we want to store things, right? So, if you look at it, the CONFERENCE_JOINED action already encapsulates what you need to store it. In fact you pass along the same `conference` object.

It's similar for CONFERENCE_ENDED.

**Comment by saghul** [comment]
> Does this look better?

Oh yeah man!

**Comment by akshitkrnagpal** [comment]
> Oh yeah man!

Cool.

File: app/features/recent-list/actionTypes.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (saghul) [reply app/features/recent-list/actionTypes.js]: > We are storing things through the recent-list reducer. Do you want me to move functionality here?

Yep

[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/recent-list/actionTypes.js]: Then we won't have a reducer there. Shouldn't this functionality be in `features/recent-list`?

Since the motive was to gain the flexibility to do more things when the conference ends (or starts) wouldn't it be better to dispatch actions of other features instead of implementing every feature's functionality here?

[the remark this answers is not in the document]
Comment (saghul) [reply app/features/recent-list/actionTypes.js]: > Then we won't have a reducer there. Shouldn't this functionality be in features/recent-list?

We can have multiple reducers for the same actions. The recent-list feature will have a reducer for the conference actions and store recent list related data.



> Since the motive was to gain the flexibility to do more things when the conference ends (or starts) wouldn't it be better to dispatch actions of other features instead of implementing every feature's functionality here?

I'm not sure I understand, can you give an example or rephrase?

[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/recent-list/actionTypes.js]: > We can have multiple reducers for the same actions. The recent-list feature will have a reducer for the conference actions and store recent list related data.

Ooh. So I'll remove the actions in recent-list and remove the middleware in the conference.

> I'm not sure I understand, can you give an example or rephrase?

I was thinking when CONFERENCE_JOINED is dispatched we'll dispatch all actions we need to at that time using middleware. But since we can have multiple reducers for the same actions, this is not needed anymore. :sweat_smile:

File: app/features/conference/components/Conference.js [hunk 10]
```diff
@@ -18,6 +18,7 @@ import {
 import config from '../../config';
 import { setEmail, setName } from '../../settings';
 
+import { notifyConferenceEnded, notifyConferenceJoined } from '../actions';
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/conference/components/Conference.js]: Sure.

File: app/features/conference/middleware.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
+// @flow
+
+import { saveCurrentConference, updateEndTime } from '../recent-list';
+
+import { CONFERENCE_JOINED, CONFERENCE_ENDED } from './actionTypes';
+
+export default (store: Object) => (next: Function) => (action: Object) => {
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/conference/middleware.js]: I'll do that.

File: app/features/recent-list/styled/index.js [hunk 12]
```diff
@@ -0,0 +1,3 @@
+export { default as ConferenceCard } from './ConferenceCard';
+export { default as RecentListContainer } from './RecentListContainer';
+export { default as TruncateText } from './TruncateText';
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/recent-list/styled/index.js]: Yeah.

File: app/features/recent-list/components/RecentList.js [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+     * Render function of component.
+     *
+     * @returns {ReactElement}
+     */
+    render() {
+        return (
+            <RecentListContainer>
+                {
+                    this.props._recentList.map(
+                        conference => this._renderConferenceCard(conference)
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/recent-list/components/RecentList.js]: Okay.

File: app/features/recent-list/components/RecentList.js [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+    /**
+     * Redux dispatch.
+     */
+    dispatch: Dispatch<*>;
+
+    /**
+     * Array of recent conferences.
+     */
+    _recentList: Array<RecentListItem>;
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/recent-list/components/RecentList.js]: Beginning.

File: app/features/conference/components/Conference.js [hunk 15]
```diff
@@ -239,11 +240,21 @@ class Conference extends Component<Props, State> {
         setupAlwaysOnTopRender(this._api);
         setupWiFiStats(iframe);
 
-        this._api.on('readyToClose', (event: Event) =>
-            this._navigateToHome(event));
+        this._api.on('readyToClose', (event: Event) => {
+            this.props.dispatch(conferenceEnded());
```
Comment (saghul) [inline app/features/conference/components/Conference.js]: we should have the conference object here too.

File: app/features/recent-list/components/RecentList.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (saghul) [reply app/features/recent-list/components/RecentList.js]: Then this needs reordering

File: app/features/recent-list/reducer.js [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+ * @param {State} state - Current reduced redux state.
+ * @param {Object} action - Action which was dispatched.
+ * @returns {State} - Updated reduced redux state.
+ */
+export default (state: State = DEFAULT_STATE, action: Object) => {
+    switch (action.type) {
+    case CONFERENCE_ENDED:
+        return {
+            ...state,
+            recentList: _updateEndtimeOfLastConference(state.recentList)
```
Comment (saghul) [inline app/features/recent-list/reducer.js]: we should use the information of the conference that is passed as a parameter (as I mentioned above). I know that at the moment this will always be the last one, but it's not unthinkable that at some point we'll start a new conference which will cause the previous one to end and I don't want us to rely on ordering.

File: app/features/recent-list/reducer.js [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+function _insertConference(
+        recentList: Array<RecentListItem>,
+        newConference: RecentListItem
+) {
+
+    // Add start time to conference.
+    newConference.startTime = Date.now();
+
+    // Remove same conference.
+    // $FlowFixMe
```
Comment (saghul) [inline app/features/recent-list/reducer.js]: what does flow complain about?

File: app/features/recent-list/reducer.js [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+/**
+ * Update the EndTime of the last conference.
+ *
+ * @param {Array<RecentListItem>} recentList - Previous recent list array.
+ * @returns {Array<RecentListItem>} - Updated recent list array.
+ */
+function _updateEndtimeOfLastConference(recentList: Array<RecentListItem>) {
+
```
Comment (saghul) [inline app/features/recent-list/reducer.js]: style: spurious empty line

File: app/features/recent-list/reducer.js (hunk 16 above)
Comment (akshitkrnagpal) [reply app/features/recent-list/reducer.js]: Even if we start a new conference the previous conference will end first. Wouldn't it?
There are no plans for two conferences at a single time?

File: app/features/recent-list/components/RecentList.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply app/features/recent-list/components/RecentList.js]: `_` variables should be above?

The whole project has `_` variables below :upside_down_face:

I'll do that in another PR Misc. ( with `em` thing )

File: app/features/recent-list/reducer.js (hunk 18 above)
Comment (akshitkrnagpal) [reply app/features/recent-list/reducer.js]: :+1:

File: app/features/recent-list/reducer.js (hunk 17 above)
Comment (akshitkrnagpal) [reply app/features/recent-list/reducer.js]: Huh... It no longer does... Thanks for catching that :smile:

File: app/features/recent-list/reducer.js (hunk 16 above)
Comment (saghul) [reply app/features/recent-list/reducer.js]: Correct on both accounts. But I don't want to rely on that kind of ordering. Explicit is better than implicit :-)

File: app/features/recent-list/components/RecentList.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (saghul) [reply app/features/recent-list/components/RecentList.js]: Ah, then this is correct, sorry for the noise.

File: app/features/recent-list/reducer.js (hunk 16 above)
Comment (akshitkrnagpal) [reply app/features/recent-list/reducer.js]: Okay. Updated.

File: app/features/recent-list/reducer.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (saghul) [reply app/features/recent-list/reducer.js]: doh, I'm an idiot
